src/random.cr

```diff
@@ -1,34 +1,179 @@
-lib C
-  RAND_MAX = 2147483647
+class Random
 
-  fun srand(seed : UInt32)
-  fun rand() : Int32
-end
+  class MT19937
+    N = 624
+    M = 397
+    MATRIX_A = 0x9908b0dfu32
+    UPPER_MASK = 0x80000000u32
+    LOWER_MASK = 0x7fffffffu32
 
-C.srand(Intrinsics.read_cycle_counter.to_u32)
+    def initialize(seeds = Array(UInt32).new(4){ Intrinsics.read_cycle_counter.to_u32 })
+      @mt = StaticArray(UInt32, 624).new(0u32)
+      @mti = N + 1
+      init_by_array(seeds)
+    end
 
-def srand(x)
-  C.srand(x.to_u32)
+    private def init_genrand(seed)
+      @mt[0] = seed & 0xffffffffu32
+      @mti = 1
+      while @mti < N
+        @mt[@mti] = (1812433253u32 * (@mt[@mti-1] ^ (@mt[@mti-1] >> 30)) + @mti) & 0xffffffffu32
+        @mti += 1
+      end
+    end
+
+    private def init_by_array(init_keys)
+      key_len = init_keys.size
+      init_genrand 19650218u32
+
+      i = 1
+      j = 0
+      k = if N > key_len
+            N
+          else
+            key_len
+          end
+
+      while k > 0
+        @mt[i] = (@mt[i] ^ ((@mt[i-1] ^ (@mt[i-1] >> 30)) * 1664525u32)) + init_keys[j] + j
+
+        i += 1
+        j += 1
+
+        if i >= N
+          @mt[0] = @mt[N-1]
+          i = 1
+        end
+
+        if j >= key_len
+          j = 0
+        end
+
+        k -= 1
+      end
+
+      k = N - 1
+
+      while k > 0
+        @mt[i] = (@mt[i] ^ ((@mt[i-1] ^ (@mt[i-1] >> 30)) * 1566083941u32)) - i
+        i += 1
+
+        if i >= N
+          @mt[0] = @mt[N-1]
+          i = 1
+        end
+
+        k -= 1
+      end
+
+      # Use to_i because substituting 0x80000000 causes SEGV
+      @mt[0] = 0x80000000u32
+    end
+
+    def next_number()
+      mag01 = [0, MATRIX_A]
```

Comment:
In terms of performance, it may be better not to use `mag01`.  `mag01[y%2]` can be replaced with `if y%2==0 then 0 else MATRIX_A`.